# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x + 2

print (add_two)

<function add_two at 0x1040b5cf0>


In [4]:
lambda x: x + 2

<function __main__.<lambda>(x)>

In [5]:
(lambda x: x + 2)(15)

17

#### Saving to a Variable

In [8]:
named_lambda = lambda z: z ** 2

#### Multiple Inputs

In [9]:
(lambda x, y, z: x * y * z)(1, 2, 3)

6

passing a function into a function

In [12]:
def apply_to_each(f, a_list):
    for element in a_list:
        f_return = f(element)
        print(f_return)


In [15]:
def get_first_letter(string):
    return string[0]



In [16]:
list_of_seasons = ['Winter', 'Spring', 'Summer', 'Fall']

In [17]:
apply_to_each(get_first_letter, list_of_seasons)

W
S
S
F


#### Passing a Lambda into a Function

In [18]:
apply_to_each(lambda string: string[-1], list_of_seasons)

r
g
r
l


In [21]:
apply_to_each(lambda x: x ** 2, [1, 2, 3, 4, 5, 6, 7, 8])

1
4
9
16
25
36
49
64


In [26]:
#RETURN A FUNCTION FROM A FUNCTION
def my_outer_function(name):
    print(f"Hello {name}")

    def my_inner_function():
        print ('This is the inner function')

        my_inner_function
print(my_outer_function)




<function my_outer_function at 0x104265f30>


In [27]:
my_outer_function('Steven')


Hello Steven


#### Returning a Lambda from a Function

In [29]:
def new_multi(num):
    return lambda x: x * num

four = new_multi(4)
print(four)

<function new_multi.<locals>.<lambda> at 0x104644790>


In [30]:
four(5)

20

#### If Statements within Lambdas

In [ ]:
def double_pos_squar_neg(number):
    if number > 0:
        return number * 2
    else:
        return number ** 2
    
    


In [31]:
(lambda num: num * 2 if num >= 0 else num ** 2)(10)

20

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [48]:
def annon():
    return (lambda num: num * num * num)
f =  annon()



(f(7))

343

In [49]:
f = lambda x: x ** 3

In [50]:
def  my_generator(number):
    i = 0
    while i < number:
        return i
    
x = my_generator(10)

print(x)

0


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [51]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [57]:
def spquared_minus_two(num):
    return num**2 - 2



numbers = [1, 2, 3, 4, 5, 6]


new_nums = []

for num in numbers:
    new_num = spquared_minus_two(num)
    new_nums.append(new_num)


print(new_nums)

[-1, 2, 7, 14, 23, 34]


In [60]:
x_values = [1, 2, 3, 4, 5, 6]
y_values = [20, 30, 40, 50, 60, 70]

def multiple (x, y):
    return x * y

list(map(multiple, x_values, y_values))

[20, 60, 120, 200, 300, 420]

#### Using Lambda's with Map

In [58]:
list(map(spquared_minus_two, numbers))

[-1, 2, 7, 14, 23, 34]

In [61]:
x_vals = [1, 2, 3, 4, 5, 6]
y_vals = [20, 30, 40, 50, 60, 70]


list(map(lambda x, y: x * y, x_vals, y_vals))


[20, 60, 120, 200, 300, 420]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [62]:
numbers = range(1,11)

# [0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]

list(map(lambda x: x*2 if x % 2 == 0 else x / 2 , numbers))

[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [64]:
names = ['Bob', 'Billy', 'Alex', 'Steve', 'Cam']

def get_name(name):
    return name[0].upper() == 'A'



a_names_map = list(map(get_name,names))
print(a_names_map)


[False, False, True, False, False]


In [67]:

a_names_filter = list(filter(get_name, names))
print(a_names_filter)

['Alex']


In [68]:
def remainder(num):
    return num % 3 

my_nums = [ x for x in range(1,10)]
print(my_nums)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [69]:
list(filter(remainder, my_nums))

[1, 2, 4, 5, 7, 8]

#### Using Lambda's with Filter()

In [74]:
list(filter(lambda names: True if names[0].upper() == 'B' else False, names))

['Bob', 'Billy']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [5]:
import statistics
from random import randint

random_nums = [randint(1, 100) for _ in range(20)]


mean = statistics.mean(random_nums)


print(mean)


list(filter(lambda num: num > mean, random_nums))


57.3


[62, 68, 70, 84, 90, 83, 83, 65, 59, 66, 94, 64]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [8]:
from functools import reduce


def add_nums(x,y):
    return x + y

results_add = reduce(add_nums, [1,2,3,4,5])
print(results_add)

15


#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [9]:
reduce(lambda x,y: x*y, [1,2,3,4,5])

120

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [16]:
import time

def capitilize(name):
    return name[0].upper() + name[1:].lower()


def full_name(first, last):
    return capitilize(first) + ' ' + capitilize(last)

def great_person(first_name, last_name, day_of_week):
    person_full_name = full_name(first_name, last_name)
    print(f"Hello {person_full_name}, welcome to the {day_of_week} day of the week!")


great_person('sTeveN', 'AdeT', 'Monday')

Hello Steven Adet, welcome to the Monday day of the week!


In [18]:
def func_one():
    print("function one started")
    time.sleep(5)
    print("function one finished")


def func_two():
    print("function two started")
    func_one()
    print("function two finished")

func_two()

function two started
function one started
function one finished
function two finished


In [21]:
def add_nums(num):
    if num <= 1:
        return num
    else:
        return num + add_nums(num - 1)
    

add_nums(10)

55

#### Writing a Factorial Function

In [22]:
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
    
print(factorial(10))

3628800


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [25]:
lengths = map(len,['Brian', 'Michael', 'Jason', 'Jeffrey', 'John'])
print (lengths)

#### Infinite Generator

In [ ]:
# bad, never create infinite loops




#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [71]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

# list(filter(lambda x: True if len(x) > 2 else False, places))


real_places = list(filter(lambda city: city.strip(), places))

print(real_places)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [53]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]




author.sort(key=lambda x: x.split()[-1])

print(author)

['Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'Victor aNisimov', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [74]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

F = list(map(lambda x: (x[0], (9/5)* x[1]+32), places))

print(F)


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [ ]:
def fib(num):
    if num <= 1:
        return 1
    else:
        return fib(num - 1) + fib(num - 2)